Feature/optimize face touch detection

wenbozzz opened the pull request:
* change `getShortestDistance` to use octree to find nearest point on face
* visualize closest point pair in pointcloud (red)
* reduce some false positive detection
![2](https://user-images.githubusercontent.com/1872200/77526233-7b626200-6e47-11ea-95e9-cbbee4d54b99.PNG)

wenbozzz committed:
use octree in getShortestDistance
use octree instead of brute force to search nearest point

wenbozzz committed:
visualize closest point in red

wenbozzz committed:
improve face touch detection
reduce false positive when hand is in front of the face

stevenguh commented on detector.js:
```diff
@@ -9,6 +9,8 @@ import { sleep } from './utils';
 import { fingerLookup, drawFacePredictions, drawHandPredictions } from './draw';
 import { BoundingBox, boxLookup } from './box';
 
+var createOctree = require('yaot'); // from https://github.com/anvaka/yaot
```
Change to `import  createOctree  from 'yaot';` to be more consistent

stevenguh commented on detector.js:
```diff
@@ -21,24 +23,38 @@ function getHandPoints(predictions) {
-            const d = Math.sqrt(Math.pow(x, 2) + Math.pow(y, 2) + Math.pow(z, 2));
-            if (shortestDistance === undefined || d < shortestDistance.d) {
-                shortestDistance = { x, y, z, d };
+function getShortestDistance(handPoints, facePoints, distance_threshold) {
```
For `distance_threshold` and other variables in the file, use camel case instead as the convention.

stevenguh replied:
Maybe we can also have a benchmark of these two and see if the new one is actually faster? Since I assume there is a bunch more memory allocation with the octree

wenbozzz replied:
reverted back to the original algorithm

stevenguh commented on detector.js:
```diff
@@ -224,36 +269,28 @@ export default class Detector {
             this.scatterGL.setPointColorer((i, selectedIndices, hoverIndex) => {
+                if (minDistance && 
+                    (i == handPoints.length + minDistance.face_point_idx || i == minDistance.hand_point_idx)) {
+                    return 'red';
```
I like this visualization

stevenguh commented on detector.js:
```diff
@@ -185,11 +201,40 @@ export default class Detector {
+        }
+
+        // regenerate hand bbox after scaling z axis
+        handBox = BoundingBox.createFromPoints(handPoints);
```
Move L206-207 after this line? Should we have bounding box that's not corresponding to the hand points?

wenbozzz replied:
fixed

stevenguh commented on detector.js:
```diff
@@ -185,11 +201,40 @@ export default class Detector {
+                    handPoints[i][2] = (handPoints[i][2] > 1) ? handPoints[i][2]*1.5 : handPoints[i][2];
+                }
+            }
+        }
```
This is very interesting, but it also made the chin detection less robust. A lot of times, chin detection seems to be harder than before. I'd like to see if there is a way to fix that.

wenbozzz replied:
see the latest change

stevenguh reviewed (changes requested):
There are some changes I'd like to see. 

1) the new shortest distance is actually faster and better
2) The chin detection seems to be worse off. Basically the overall detection is reduced hence also the false positive.
3) Change variables to camelCase
4) I really like the closest point visualization. No change here.

wenbozzz committed:
format code to follow convension

stevenguh commented:
Just did a simple benchmark on `getShortestDistance` implementation with both bruce force and with Octree.
The function takes around 0.2 ms with Octree and 0.05 ms with bruce force. Unless we found a better use case like multi-hand and multi-face for Octree, I think we should keep the bruce force implementation for now.

wenbozzz commented:
Yeah, on my laptop it was about the same result 😂
I'm reverting the algorithm back
![time_brute_force](https://user-images.githubusercontent.com/1872200/77800904-06ed1600-7035-11ea-9092-8bf7b8324cea.PNG)
![time_octree](https://user-images.githubusercontent.com/1872200/77800908-08b6d980-7035-11ea-9ed1-46e5e0b447fa.PNG)

wenbozzz committed:
undo e474f7bd5121680fdcd2113b077240fea7169b32
remove using octree since it performed even worse
brute force took 0.03 ms on avg
octree took 1.5 ms on avg

wenbozzz committed:
fix handBoxPoints
compute handBoxPoints after hand points z axis rescaled

wenbozzz committed:
tune parameter for false positive reduction

stevenguh commented:
I still get some performance reduction on the chin, but let's try it.

wenbozzz commented:
> I still get some performance reduction on the chin, but let's try it.

maybe we could add a UI element to tune the parameter like `timeout`?

stevenguh commented:
We can tune it for sure but just FYI the lower the time out the more resources we are going to use. We ideally want this app to be able to run in the background while people work on their computer. Cody changes the timeout to 300 ms on his UI code. we will have to see if that works well for most people.
